src/core/model/Monitor.h

```diff
@@ -0,0 +1,44 @@
+#include <cassert>
+#include <mutex>
+
+template<class T>
+class Monitor
+{
+public:
+    template<typename ...Args>
+    Monitor(Args&&... args);
+
+    struct LockedMonitor
+    {
+        LockedMonitor(Monitor* monitor);
+        T* operator->();
+        void ReplaceModel(T model);
+        private:
+            Monitor* mon;
+            std::unique_lock<std::mutex> lock;
+    };
+
+    struct TryLockedMonitor
+    {
+        TryLockedMonitor(Monitor* monitor);
+        T* operator->();
+        void ReplaceModel(T model);
+        bool lockAcquired;
+        private:
+            Monitor* mon = nullptr;
+            std::unique_lock<std::mutex> lock;
+    };
+
+
+    LockedMonitor operator->();
+
+    LockedMonitor lock();
+
+    TryLockedMonitor tryLock();
+
+    T& getUnsafeAccess();
```

Comment:
We actually don't want such a function, since access should be guarded. But since it helps while refactoring, mark it `[[deprecated]]`:
```suggestion
    [[deprecated]] T& getUnsafeAccess();
```